### PR TITLE
[Pallas] Support ab for flash_attention

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull request adds ab support for flash_attention which is a custom mask for attention weight.

Test Plan:
PJRT_DEVICE=TPU python test/test_pallas.py -v -k test_flash_attention_ab

### CODE REVIEW

**Comment by ZhiyuLi-goog** [comment]
Thank you @alanwaketan!

File: test/test_pallas.py [hunk 1]
```diff
@@ -920,6 +922,106 @@ def test_flash_attention_sm_scale_backward(self):
       self.assertTrue(torch.allclose(i[0].grad.cpu(), i[1].cpu(), atol=1e-05))
     jax.config.update("jax_default_matmul_precision", "default")
 
+  @unittest.skipIf(xr.device_type() != 'TPU' or tpu.version() < 3,
+                   "This test only works on TPUv3+.")
+  def test_flash_attention_ab(self):
+    jax.config.update("jax_default_matmul_precision", "highest")
```
Comment (JackCaoG) [inline test/test_pallas.py:928]: lol we should really make a context manager that take cares of this in this test.

**Review by JackCaoG — APPROVED** [review]
Can you rerun the CI? TPUCI was not enabled so the test was not run.

**Comment by JackCaoG** [comment]
actually let me just trigger it...

**Comment by alanwaketan** [comment]
Thanks, Jack.
